gazebo_plugins/src/gazebo_ros_ackermann_drive.cpp

```diff
@@ -0,0 +1,592 @@
+  auto linear_vel = joints_[REAR_RIGHT]->GetVelocity(0);
+  auto target_linear = ignition::math::clamp(target_linear_, -max_speed_, max_speed_);
+  double linear_diff = linear_vel - target_linear / wheel_radius_;
+  double linear_cmd = pid_[REAR_RIGHT].Update(linear_diff, seconds_since_last_update);
+
+  auto target_rot = target_rot_ * copysign(1.0, target_linear_);
+  target_rot = ignition::math::clamp(target_rot, -max_steer_, max_steer_);
+
+  double tanSteer = tan(target_rot);
+
+  auto target_left_steering =
+    atan2(tanSteer, 1.0 - wheel_separation_ / 2.0 / wheel_base_ * tanSteer);
+  auto target_right_steering =
+    atan2(tanSteer, 1.0 + wheel_separation_ / 2.0 / wheel_base_ * tanSteer);
+
+  auto left_steering_angle = joints_[STEER_LEFT]->Position(2);
+  auto right_steering_angle = joints_[STEER_RIGHT]->Position(2);
+
+  double left_steering_diff = left_steering_angle - target_left_steering;
+  double left_steering_cmd = pid_[FRONT_LEFT].Update(left_steering_diff, seconds_since_last_update);
+
+  double right_steering_diff = right_steering_angle - target_right_steering;
+  double right_steering_cmd =
+    pid_[FRONT_RIGHT].Update(right_steering_diff, seconds_since_last_update);
+
+  auto steer_wheel_angle = (left_steering_angle + right_steering_angle) * 0.5 / steering_ratio_;
+
+  joints_[STEER_LEFT]->SetForce(0, left_steering_cmd);
+  joints_[STEER_RIGHT]->SetForce(0, right_steering_cmd);
+  joints_[STEER_WHEEL]->SetPosition(0, steer_wheel_angle);
+  joints_[REAR_RIGHT]->SetForce(0, linear_cmd);
+  joints_[REAR_LEFT]->SetForce(0, linear_cmd);
+
+  last_update_time_ = _info.simTime;
+}
+
+void GazeboRosAckermannDrivePrivate::OnCmdVel(const geometry_msgs::msg::Twist::SharedPtr _msg)
+{
+  std::lock_guard<std::mutex> scoped_lock(lock_);
+  target_linear_ = _msg->linear.x;
+  target_rot_ = _msg->angular.z;
+}
+
+double GazeboRosAckermannDrivePrivate::CollisionRadius(const gazebo::physics::CollisionPtr & _coll)
+{
+  if (!_coll || !(_coll->GetShape())) {
+    return 0;
+  }
+  if (_coll->GetShape()->HasType(gazebo::physics::Base::CYLINDER_SHAPE)) {
+    gazebo::physics::CylinderShape * cyl =
+      dynamic_cast<gazebo::physics::CylinderShape *>(_coll->GetShape().get());
+    return cyl->GetRadius();
+  } else if (_coll->GetShape()->HasType(gazebo::physics::Base::SPHERE_SHAPE)) {
+    gazebo::physics::SphereShape * sph =
+      dynamic_cast<gazebo::physics::SphereShape *>(_coll->GetShape().get());
+    return sph->GetRadius();
+  }
+  return 0;
+}
+
+void GazeboRosAckermannDrivePrivate::UpdateOdometryWorld()
+{
+  auto prev_x = odom_.pose.pose.position.x;
+  auto prev_y = odom_.pose.pose.position.y;
+
+  auto pose = model_->WorldPose();
+  odom_.pose.pose.position = gazebo_ros::Convert<geometry_msgs::msg::Point>(pose.Pos());
+  odom_.pose.pose.orientation = gazebo_ros::Convert<geometry_msgs::msg::Quaternion>(pose.Rot());
+
+  distance_.data += hypot(prev_x - odom_.pose.pose.position.x, prev_y - odom_.pose.pose.position.y);
```

Comment:
This seems to be drifting too fast. I let the Prius sit for 30 sec and this is publishing a distance of 0.19 m. But if I check the model pose on the GUI, it's just -0.09 on X. Maybe we need to double-check the math?

Reply:
I don't think there is a problem with the math. Actually the problem is that the model pose seems to be oscillating in the Y direction by very small amount when static. So over time the error accumulates.